client/go/outline/device.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package outline
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+
+	perrs "github.com/Jigsaw-Code/outline-apps/client/go/outline/platerrors"
+	"github.com/Jigsaw-Code/outline-apps/client/go/outline/vpn"
+	"github.com/Jigsaw-Code/outline-sdk/network"
+	"github.com/Jigsaw-Code/outline-sdk/network/dnstruncate"
+	"github.com/Jigsaw-Code/outline-sdk/network/lwip2transport"
+)
+
+// Device is an IPDevice that connects to a remote Outline server.
+// It also implements the vpn.ProxyDevice interface.
+type Device struct {
+	network.IPDevice
+
+	c           *Client
+	pkt         network.DelegatePacketProxy
+	supportsUDP bool
+
+	remote, fallback network.PacketProxy
+}
+
+var _ vpn.ProxyDevice = (*Device)(nil)
+
+// NewDevice creates a new [Device] using the given [Client].
+func NewDevice(c *Client) (*Device, error) {
+	if c == nil {
+		return nil, errors.New("Client must be provided")
+	}
+	return &Device{c: c}, nil
+}
+
+// SupportsUDP returns true if the the Outline server forwards UDP traffic.
+// This value will be refreshed after Connect or RefreshConnectivity.
+func (d *Device) SupportsUDP() bool {
+	return d.supportsUDP
+}
+
+// Connect tries to connect to the Outline server.
+func (d *Device) Connect(ctx context.Context) (err error) {
```

Comment:
Can we simplify and have a factory function that connects and creates the `Device`?

Instead of `NewDevice(client)` and `Connect(ctx)`, have a single `Connect(ctx, client) Device`.

Some of these methods only make sense after the connection (like SupportsUDP, RefreshConnectivity), so it will be cleaner to only have the Device when it's already connected.

I believe you can get rid of the `vpn.ProxyDevice` interface if you move this remote `Device` to the `vpn` package.

Reply:
Done.